### PR TITLE
Index ast updates

### PR DESCRIPTION
**Pull request by jeffreylovitz**
This PR structures indexing operations in a separate clause file, matching the organization of the other elements of a user query.

It updates the AST_Query struct and New_AST_Query function to handle these additions.

Within the parser module, this can be considered a complete change. It differs somewhat from the earlier implementation of indexing, so CREATE INDEX queries will fail until the associated PRs are also merged.

### CODE REVIEW

File: src/parser/clauses/index.c [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#include "./index.h"
+#include "../ast_common.h"
+
+AST_IndexNode* AST_IndexOp(const char *label, const char *property, AST_IndexOpType optype) {
+  AST_IndexNode *indexOp = malloc(sizeof(AST_IndexNode));
+  indexOp->target.label = label;
+  indexOp->target.property = property;
```
Comment (jeffreylovitz) [inline src/parser/clauses/index.c]: In my current testing, it has not proven necessary to `strdup` these inputs, but I could be overlooking something? Let me know if I should change this!

Comment (swilly22) [reply src/parser/clauses/index.c]: You're right, we don't have to duplicate these.

File: src/parser/grammar.y [hunk 2]
```diff
@@ -36,33 +36,49 @@
...
-	A = New_AST_Query(NULL, NULL, B, NULL, NULL, NULL, NULL, NULL, NULL);
+	A = New_AST_Query(NULL, NULL, B, NULL, NULL, NULL, NULL, NULL, NULL, NULL);
 }
 
 //expr(A) ::= mergeClause(B). {
-//	A = New_AST_Query(NULL, NULL, NULL, B, NULL, NULL, NULL, NULL, NULL);
+//	A = New_AST_Query(NULL, NULL, NULL, B, NULL, NULL, NULL, NULL, NULL, NULL);
 //}
 
+expr(A) ::= CREATE INDEX ON indexLabel(B) indexProp(C) . {
```
Comment (swilly22) [inline src/parser/grammar.y]: Maybe consider merging the two index rules into one, and add a new rule for parsing the index CREATE/DROP operations:
expr(A) ::= indexOp(B) INDEX ON indexLabel(C) indexProp(D) .

File: src/parser/grammar.y [hunk 3]
```diff
@@ -36,33 +36,50 @@
...
 //}
 
+expr(A) ::= indexOpToken(B) INDEX ON indexLabel(C) indexProp(D) . {
+	A = New_AST_Query(NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, AST_IndexOp(C.strval, D.strval, B));
+}
+
+%type indexOpToken { AST_IndexOpType }
+
+indexOpToken(A) ::= CREATE . { A = CREATE_INDEX; }
+indexOpToken(A) ::= DROP . { A = DROP_INDEX; }
```
Comment (jeffreylovitz) [inline src/parser/grammar.y]: I agree with your idea of simplifying the index rules - does this seem like an appropriate solution?

I couldn't think of how to handle CREATE and DROP in a single rule, so instead took this approach (following the style of the `relation(A)` rules used for Filters).

If you have a simpler solution, I'd be happy to implement it!

File: src/parser/clauses/index.h [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  const char *label;
+  const char *property;
+} IndexTarget;
+
+typedef struct {
+  IndexTarget target;
+  AST_IndexOpType operation;
+} AST_IndexNode;
+
+AST_IndexNode* AST_IndexOp(const char *label, const char *property, AST_IndexOpType optype);
```
Comment (swilly22) [inline src/parser/clauses/index.h]: Please use a function name similar to the other clauses, see create clause for example.

File: src/parser/grammar.y [hunk 5]
```diff
@@ -36,33 +36,50 @@
...
-	A = New_AST_Query(NULL, NULL, B, NULL, NULL, NULL, NULL, NULL, NULL);
+	A = New_AST_Query(NULL, NULL, B, NULL, NULL, NULL, NULL, NULL, NULL, NULL);
 }
 
 //expr(A) ::= mergeClause(B). {
-//	A = New_AST_Query(NULL, NULL, NULL, B, NULL, NULL, NULL, NULL, NULL);
+//	A = New_AST_Query(NULL, NULL, NULL, B, NULL, NULL, NULL, NULL, NULL, NULL);
 //}
 
+expr(A) ::= indexOpToken(B) INDEX ON indexLabel(C) indexProp(D) . {
```
Comment (swilly22) [inline src/parser/grammar.y]: Don't include the entire rule for parsing an index request directly within expr,
follow the same schema other expression follow, see line 58 for example.

File: src/parser/grammar.y (hunk 3 above)
Comment (swilly22) [reply src/parser/grammar.y]: This is perfect.

File: src/parser/grammar.y [hunk 6]
```diff
@@ -92,6 +96,25 @@ createClause(A) ::= CREATE chains(B). {
...
+indexClause(A) ::= indexOpToken(B) INDEX ON indexLabel(C) indexProp(D) . {
+  A = New_AST_IndexNode(C.strval, D.strval, B);
+}
+
+%type indexOpToken { AST_IndexOpType }
+
+indexOpToken(A) ::= CREATE . { A = CREATE_INDEX; }
+indexOpToken(A) ::= DROP . { A = DROP_INDEX; }
+
+indexLabel(A) ::= COLON UQSTRING(B) . {
```
Comment (swilly22) [inline src/parser/grammar.y]: Don't we need to declare a type for indexLabel?

**Pull request merged**